### PR TITLE
Exporter for EclipseLink library

### PR DESCRIPTION
**Pull request by VitaNuova**
This exporter follows the same structure as the Hibernate exporter, but exports the metrics available in EclipseLink. I had to write it for my project and thought I might try to contribute it. @brian-brazil I would be happy about any feedback.

### CODE REVIEW

File: simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java [hunk 1]
```diff
@@ -0,0 +1,744 @@
...
+      throw new IllegalStateException("You must register at least one Session.");
+    }
+    return super.register(registry);
+  }
+
+  private List<MetricFamilySamples> getCounters() {
+    return Arrays.<MetricFamilySamples>asList(
+        createCounter(
+            "eclipselink_uow_commits_total",
+            "Total number of unit of work commits",
```
Comment (brian-brazil) [inline simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java]: Is uow a common term?

File: simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java [hunk 2]
```diff
@@ -0,0 +1,744 @@
...
+                if (value != null) {
+                  return ((Integer) value).doubleValue();
+                }
+                return 0;
+              }
+            }
+        ),
+        createCounter(
+            "eclipselink_cache_misses_total",
+            "Total number of cache misses",
```
Comment (brian-brazil) [inline simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java]: It's best to have cache requests as a metric too, it makes it easier to calculate the hit ratio.

File: simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java [hunk 3]
```diff
@@ -0,0 +1,744 @@
...
+                if (value != null) {
+                  return ((Integer) value).doubleValue();
+                }
+                return 0;
+              }
+            }
+        ),
+        createCounter(
+            "eclipselink_change_sets_processed_total",
+            "Total number of processed change sets",
```
Comment (brian-brazil) [inline simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java]: Are this one and the next one successes and failures? Failures and total is a bit easier to work with

File: simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java [hunk 4]
```diff
@@ -0,0 +1,744 @@
...
+                Object value = metrics.get(SessionProfiler.OptimisticLockException);
+                if (value != null) {
+                  return ((Integer) value).doubleValue();
+                }
+                return 0;
+              }
+            }
+        ),
+        createCounter(
+            "eclipselink_rcm_received_total",
```
Comment (brian-brazil) [inline simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java]: What's rcm?

File: simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java [hunk 5]
```diff
@@ -0,0 +1,744 @@
...
+              }
+            }
+        )
+    );
+  }
+
+  private List<MetricFamilySamples> getTimers() {
+    return Arrays.<MetricFamilySamples>asList(
+        createGauge(
+            "eclipselink_uow_commits_seconds",
```
Comment (brian-brazil) [inline simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java]: These all sound like counters

Comment (VitaNuova) [reply simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java]: They all measure durations of particular operations in seconds, not sure how they can be counters. As durations, they can arbitrarily go up and down.

File: simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java (hunk 1 above)
Comment (VitaNuova) [reply simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java]: Afaik, Unit of Work is a common term (see also https://stackoverflow.com/questions/3532258/unit-of-work-pattern-and-history). However, I should probably avoid the abbreviation in the metric name.

File: simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java (hunk 5 above)
Comment (brian-brazil) [reply simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java]: In that case see if you can get to the underlying data, and create a Summary from that

File: simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java (hunk 2 above)
Comment (VitaNuova) [reply simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java]: Agree, but the metric as such was not present in EclipseLink statistics. I can add cache requests as the sum of cache hits and cache misses.

File: simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java (hunk 4 above)
Comment (VitaNuova) [reply simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java]: RemoteCommandManager (https://www.eclipse.org/eclipselink/api/2.2/org/eclipse/persistence/sessions/coordination/RemoteCommandManager.html). I can also spell out the whole thing in the metric name.

**Comment by VitaNuova** [comment]
Thank you for the comments :)

File: simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java (hunk 5 above)
Comment (VitaNuova) [reply simpleclient_eclipselink/src/main/java/io/prometheus/client/eclipselink/EclipseLinkStatisticsCollector.java]: I see your point, but EclipseLink API returns one value at a time which represents the cumulative duration of operation. Unfortunately, I don't see how I can get enough data for a summary or a histrogram in one scrape.

**Pull request closed without merging**